fairseq/training.py

```diff
@@ -0,0 +1,490 @@
+import logging
+import os
+import subprocess as sp
+import yaml
+import sys
+import copy
+import collections.abc
+import re
+from typing import Dict, Any
+
+from sisyphus import Task, Job, tk
+from sisyphus.hash import sis_hash_helper
+
+import i6_core.util as util
+
+
+class FairseqHydraConfig:
+    """
+    An object that manages a Fairseq hydra config.
+    """
+
+    def __init__(
+        self,
+        config_dict: Dict[str, Any],
+        post_config_dict: Optional[Dict[str, Any]] = None,
+        *,
+        package_name: str = "",
+    ):
+        """
+        :param config_dict: Contains the information which is needed for fairseq-hydra-train. Will be converted and dumped into a .yaml
+        :param dict post_config_dict: dictionary of the FairseqHydraConfig config variables that are not hashed
+        :param package_name: The @package directory that is required to be added to the top of Hydra config, for example "# @package _group_"
+        """
+        assert isinstance(config_dict, dict)
+        assert isinstance(post_config_dict, dict) or post_config_dict is None
+        self.config_dict = config_dict
+        self.post_config_dict = post_config_dict if post_config_dict is not None else {}
+        self.package_name = package_name
+        self.check_consistency()
+
+    def write(self, path: str):
+        config_dict = self.config_dict.copy()
+        config_dict = util.update_nested_dict(config_dict, self.post_config_dict)
+
+        # recursively go through config dictionary to get all sisyphus paths inplace
+        config_dict = util.instanciate_delayed(config_dict)
+
+        config_yaml = yaml.dump(config_dict)
+        if self.package_name != "":
+            config_yaml = f"# @package {self.package_name}\n" + config_yaml
+        with open(path, "w") as file:
+            file.write(config_yaml)
+
+    def update(self, other):
+        """
+        updates a FairseqHydraConfig with another FairseqHydraConfig:
+          * config_dict, post_config_dict use dict.update
+        :param FairseqHydraConfig other:
+        """
+        self.config_dict = util.update_nested_dict(self.config_dict, other.config_dict)
+        self.post_config_dict = util.update_nested_dict(
+            self.post_config_dict, other.post_config_dict
+        )
+        self.package_name = other.package_name
```

Comment:
```suggestion
        if other.package_name != "":
            self.package_name = other.package_name
```
proposal: update only if new name is not empty.